python/perf-kernels/rmsnorm.py

```diff
@@ -37,28 +37,87 @@ def get_autotune_config():
         return get_hip_autotune_config()
 
 
+# accumulate sum of squares for a row in a blocked manner
+@triton.jit
+def accumulate_sum_squares(input_ptr, input_row_stride, n_cols, BLOCK_SIZE, row_idx):
+    col_offsets = tl.arange(0, BLOCK_SIZE)
+    sum_squares = tl.zeros([1], dtype=tl.float32)
+    row_input_ptr = input_ptr + row_idx * input_row_stride
+
+    n_cols_blks = tl.cdiv(n_cols, BLOCK_SIZE) - 1
+    for start in range(0, n_cols_blks * BLOCK_SIZE, BLOCK_SIZE):
+        cols = start + col_offsets
+        input_ptrs = row_input_ptr + cols
+        input_ptrs = tl.multiple_of(input_ptrs, (16, ))
+        x = tl.load(input_ptrs)
+        sum_squares += tl.sum(x * x, axis=0)
+
+    # loop peeling for mask
+    cols = n_cols_blks * BLOCK_SIZE + col_offsets
+    mask = cols < n_cols
+    input_ptrs = row_input_ptr + cols
+    input_ptrs = tl.multiple_of(input_ptrs, (16, ))
+    x = tl.load(input_ptrs, mask=mask, other=0.0, cache_modifier=".cg")
+    sum_squares += tl.sum(x * x, axis=0)
+
+    return sum_squares
+
+
+# apply normalization to each block of the row
+@triton.jit
+def apply_normalization(input_ptr, output_ptr, g_ptr, input_row_stride, output_row_stride, n_cols, norm_factor,
+                        BLOCK_SIZE, row_idx):
+    col_offsets = tl.arange(0, BLOCK_SIZE)
+    row_input_ptr = input_ptr + row_idx * input_row_stride
+    row_output_ptr = output_ptr + row_idx * output_row_stride
+
+    for start in range(0, n_cols, BLOCK_SIZE):
+        cols = start + col_offsets
+        mask = cols < n_cols
+        input_ptrs = row_input_ptr + cols
+        input_ptrs = tl.multiple_of(input_ptrs, (16, ))
+        g_ptrs = g_ptr + cols
+        output_ptrs = row_output_ptr + cols
+        x = tl.load(input_ptrs, mask=mask, other=0.0, cache_modifier=".cg")
```

Comment:
Have you tried peeling the last iter? is it worth trying? Can you add a TODO to try that as part of your next PR?

Reply:
the loop peeling only applied for blocked version, not this one. sure, I will try that